File: 283-move-zeroes/move-zeroes.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int l = 0, r = 1;
        while (r < nums.size())
        {
            if (nums[l] == 0 && nums[r] != 0)
            {
                // swap using XOR handy trick, without temp variable i.e without extra space
                nums[l] ^= nums[r]; // Step 1
                nums[r] ^= nums[l]; // Step 2
                nums[l] ^= nums[r]; // Step 3
                l++;
            }
            else if (nums[l] != 0)
                l++;
            r++;
        }
    }
};

File: 283-move-zeroes/move-zeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        l, r = 0, 1
        while r < len(nums):
            if nums[l] != 0:
                l += 1
            elif nums[l] == 0 and nums[r]:
                nums[l], nums[r] = nums[r], nums[l] # swap
                l += 1
            r += 1
        